units/host_setup_script.go

```diff
@@ -0,0 +1,132 @@
+package units
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/host"
+	"github.com/evergreen-ci/utility"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/dependency"
+	"github.com/mongodb/amboy/job"
+	"github.com/mongodb/amboy/registry"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+)
+
+const (
+	hostSetupScriptJobName = "host-setup-script"
+	setupScriptRetryLimit  = 5
+)
+
+func init() {
+	registry.AddJobType(hostSetupScriptJobName, func() amboy.Job { return makeHostSetupScriptJob() })
+}
+
+type hostSetupScriptJob struct {
+	HostID         string `bson:"host_id" json:"host_id" yaml:"host_id"`
+	job.Base       `bson:"job_base" json:"job_base" yaml:"job_base"`
+	CurrentAttempt int `bson:"current_attempt" json:"current_attempt" yaml:"current_attempt"`
+
+	host *host.Host
+	env  evergreen.Environment
+}
+
+func makeHostSetupScriptJob() *hostSetupScriptJob {
+	j := &hostSetupScriptJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    hostSetupScriptJobName,
+				Version: 0,
+			},
+		},
+	}
+	j.SetDependency(dependency.NewAlways())
+	return j
+}
+
+// NewHostSetupScriptJob creates a job that executes the setup script after task data is loaded onto the host.
+func NewHostSetupScriptJob(env evergreen.Environment, h *host.Host, attempt int) amboy.Job {
+	j := makeHostSetupScriptJob()
+	j.env = env
+	j.host = h
+	j.HostID = h.Id
+	j.CurrentAttempt = attempt
+	j.SetPriority(1)
+	ts := utility.RoundPartOfHour(2).Format(TSFormat)
+	j.SetID(fmt.Sprintf("%s.%s.%s.%d", hostSetupScriptJobName, j.HostID, ts, j.CurrentAttempt))
+	return j
+}
+
+func (j *hostSetupScriptJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+	defer func() {
+		if j.HasErrors() {
+			if err := j.tryRequeue(ctx); err != nil {
```

Comment:
For the first point, I think logging but not returning the error from running the setup script is sufficient. In this case, the job won't retry. How can we check for the latter? Does that require a crons to find hosts where a script exists but the flag isn't set / enqueues a job for it?

Reply:
1. SGTM

2. Typically, all jobs that rely on re-queueing themselves have mostly assumed that there will be some cron job to re-enqueue it eventually based on some DB field if it fails to re-enqueue itself - I believe most of the re-enqueueing logic is to reduce the waiting time in case the cron job doesn't not run frequently enough.

I feel like it might be fine to not hande context cancellation depending on and just saying we don't provide guarantees that your script will run, in which case you should just run it yourself when you SSH into the host. This is mostly a question of how reliable do we want to make this script - if the job gives no promises of successfully running the script in an edge case (i.e. app server restart), then this seems fine as is.